### PR TITLE
Add NuGet Server Implementation Guide

### PR DESCRIPTION
**Pull request by zivkan**
Fixes: https://github.com/NuGet/docs.microsoft.com-nuget/issues/3059

### CODE REVIEW

File: docs/api/implementation-guide.md
Comment (zivkan) [inline docs/api/implementation-guide.md:41]: @theot12 looking at the search resource docs, it doesn't contain deprecation or vulnerability info. Was this an oversight or is it intentional?

@donnie-msft @martinrrm I'm not sure what the status of PM UI no longer hitting registration to get deprecation and/or vulnerability info is, when the search endpoint returns the data. I'm not even sure if either of you are the right person to talk to about the client side of this feature.  Should we be documenting anything related to the search resource here, regarding deprecation and vulnerability information?

File: docs/api/implementation-guide.md
Comment (zivkan) [inline docs/api/implementation-guide.md:47]: @joelverhagen is this right?  Does nuget.org create a new catalog leaf when deprecation and/or vulnerability info changes?

File: docs/api/implementation-guide.md
Comment (zivkan) [inline docs/api/implementation-guide.md:25]: This is very wordy, and I'm not sure it's clear.  What do you all think if I replace it with this table?

|URL|Will PreAuthenticate?|
|--|--|
|https://pkgs.contoso.com/nuget/v3/service/index.json|n/a, this is the service index|
|https://pkgs.contoso.com/nuget/v3/search|no, not in the same directory, or subdirectory, of the service index.|
|https://pkgs.contoso.com/nuget/v3/registration/|no, not in the same directory, or subdirectory, of the service index.|
|https://pkgs.contoso.com/nuget/v3/service/search|Yes, in the same directory as the service index.|
|https://pkgs.contoso.com/nuget/v3/service/registration|yes, in a subdirectory of the service index.|
|https://pkgs.contoso.com/nuget/v3/{guid}/search|Only if the service index URL redirected to https://pkgs.contoso.com/nuget/v3/{guid}/index.json. Otherwise, even if the GUID is the canonical name of the feed and all other names are aliases, since it's not the URL that `HttpClientHandler` used in its credential cache, these requests will not be pre-authenticated.|

Comment (nkolev92) [reply docs/api/implementation-guide.md:25]: I like the able better.

**Comment by zivkan** [comment]
Hey @loic-sharma !  I know you've implemented a NuGet server from scratch. If you have a bit of time to review and provide feedback, that'd be great.

File: docs/api/implementation-guide.md
Comment (martinrrm) [reply docs/api/implementation-guide.md:55]: From what I know, we are still using registration API to get vulnerabilities and deprecation information for all the versions. Search returns this information for the latest version only.

File: docs/api/implementation-guide.md
Comment (loic-sharma) [inline docs/api/implementation-guide.md:19]:
```suggestion
As of NuGet 6.7 (Visual Studio & MSBuild 17.7, and .NET SDK 7.0.400), NuGet uses [.NET's `HttpClientHandler.PreAuthenticate`](/dotnet/api/system.net.http.httpclienthandler.preauthenticate), which only works when subsequent URLs are in the same virtual directory, or a subdirectory, of a URL that has previously been authenticated.
```

File: docs/api/implementation-guide.md
Comment (loic-sharma) [reply docs/api/implementation-guide.md:25]: FWIW, I found the text perfectly readable but a table does add visual variety. Both seem like good options :)

File: docs/api/implementation-guide.md
Comment (loic-sharma) [inline docs/api/implementation-guide.md:27]: Is this a strict requirement? Or is this a recommendation to improve performance?

The line above indicated this was an optional performance improvement:

> This will dramatically reduce the number of unauthenticated HTTP requests sent to the server, and therefore will reduce your server workload.

File: docs/api/implementation-guide.md
Comment (loic-sharma) [inline docs/api/implementation-guide.md:31]: Should this be `are`? Not sure 😅

```suggestion
Package [icons](../reference/nuspec.md#icon), [license](../reference/nuspec.md#license), and [readme](../reference/nuspec.md#readme) files can be (and are recommended to be) embedded in the package.
```

File: docs/api/implementation-guide.md
Comment (loic-sharma) [inline docs/api/implementation-guide.md:15]: One may misunderstand "private feed" as a "custom feed that's not nuget.org" instead of "feed that requires authentication". To clear up this confusion, I'd consider starting this paragraph with something like "For feeds that require authentication, ..."

File: docs/api/implementation-guide.md
Comment (zivkan) [reply docs/api/implementation-guide.md:27]: It's just for performance. I've changed the wording, so hopefully it's clearer now.

File: docs/api/implementation-guide.md
Comment (zivkan) [reply docs/api/implementation-guide.md:31]: English is hard. Can we write docs in a formal mathematical language instead?  `&& Recommended()`

File: docs/api/implementation-guide.md
Comment (loic-sharma) [inline docs/api/implementation-guide.md:59]: Super nitpick since the field is deprecated:

```suggestion
The `<owners>` field was intended to denote who published the package on a repository, and therefore who should be contacted in case of packing issues or questions.
```

File: docs/api/implementation-guide.md
Comment (loic-sharma) [inline docs/api/implementation-guide.md:76]:
```suggestion
For the `VulnerabilityInfo` resource, nuget.org provides vulnerability data for all GitHub reviewed advisories from the [GitHub Advisories database](https://github.com/advisories), which includes packages that are not hosted on nuget.org.
```

**Comment by loic-sharma** [comment]
Is the goal of this document to make it easier for server implementors to stay up-to-date with the best practices? Will we keep adding more information to this document over time?

If yes, it may be useful to structure this document to be time-oriented like a changelog. This would make it easier for server implementors to check what's new without having to read the entire document. That said, the document's length is very reasonable so I don't feel strongly about this!

Otherwise, this doc looks great! Thanks for writing this!

File: docs/api/implementation-guide.md
Comment (loic-sharma) [reply docs/api/implementation-guide.md]: (Apparently I can't resolve my own comments?) Looks good!

File: docs/api/implementation-guide.md
Comment (joelverhagen) [reply docs/api/implementation-guide.md]: Why not put the resource name instead of the URL in the first column? (and link the resource name to the respective doc)

File: docs/api/implementation-guide.md
Comment (joelverhagen) [reply docs/api/implementation-guide.md:74]: Yes, it does. You can consider linking to this how-to in addition to the catalog resource: https://learn.microsoft.com/en-us/nuget/guides/api/query-for-all-published-packages

File: docs/api/implementation-guide.md
Comment (zivkan) [reply docs/api/implementation-guide.md]: @joelverhagen I don't understand the suggestion. Can you show an example?

I'm not sure if this answers your question, but one of the issues is that github limits the width it renders markdown, so if there are too many columns, or if there's too much text, everything gets wrapped. For URLs in particular, this gets harder to read.

**Review by nkolev92 — APPROVED** [review]

File: docs/api/implementation-guide.md
Comment (joelverhagen) [inline docs/api/implementation-guide.md:12]: This is a pretty strong statement. The shapes are quite different and in practice data manipulation occurs even for the fields that overlap between the XML .nuspec and the JSON search response. For example, the tags returned is an array, version normalization occurs (.nuspec is the input, but is not guaranteed to be normalize), the version list in search results is not available in the .nuspec (a nuspec is a version specific thing), etc etc

Perhaps I am missing some intent of the document but it reads a bit oddly to me.

What about an opening that starts a bit broader, excludes the nuance of history, while still alluding to the fact the server data mostly comes from package metadata. How about:

> In some cases, you may want to implement your own NuGet package feed. [Many existing implementations](https://learn.microsoft.com/en-us/nuget/hosting-packages/overview) exist which allow you to host your own feed in a variety way, but the protocol between the official NuGet client software and a package feed is documented allowing you to build your own feed implementation from scratch.
>
> The protocol does evolve over time and this guide is aimed at those that wish to or already have implemented a NuGet package server.
>
> Since the initial release of the NuGet V3 protocol in 2015, NuGet has evolved to provide developers with a richer experience, and this requires package repositories to do additional work in order to provide the additional value their package consumers, beyond simply exacting metadata from hosted packages and returning the metadata in various forms. For example, the [search](./search-query-service-resource.md) and [package metadata](./registration-base-url-resource.md) endpoints contain more than just metadata found in the `nupkg`'s [`nuspec` file](../reference/nuspec.md).
> 
> Note that this guide is focused on the NuGet V3 protocol since the V2 protocol is essentially undocumented and, since 2015, the recommended protocol for NuGet client and server communication is the V3 protocol. For more information read about [protocol versioning](https://learn.microsoft.com/en-us/nuget/api/overview#versioning).

Feel free to scrap my wording but the current intro focuses on one very specific confusion point which I think is too specific for such a broad guide.

File: docs/api/implementation-guide.md
Comment (joelverhagen) [inline docs/api/implementation-guide.md:19]: I think this should say "NuGet client version" since we don't really have a protocol version aside from "V3" and some resource version strings which are not guaranteed to be exactly matching the respective client version that first shipped the feature.

I think a date column would be best for the first column. That would be easiest for people to remember or refer against their own effort (e.g. "oh yeah I first implemented my Go-based server in 2021 based on Git commits, so I should look at the stuff after 2021").

File: docs/api/implementation-guide.md
Comment (joelverhagen) [inline docs/api/implementation-guide.md:34]:
```suggestion
This was explained in [a blog post from 2013](https://devblogs.microsoft.com/nuget/managing-package-owners/), so the `<owners>` field is considered deprecated in the `.nuspec` file.
```

File: docs/api/implementation-guide.md
Comment (joelverhagen) [inline docs/api/implementation-guide.md:47]: This is not strictly true per protocol. Several feeds allow mutability today and it's more that client caches things in a couple tricky ways making package mutability problematic or at least harder to navigate for the package consumer. 

I'm okay with keeping it simple in this guide since suggesting they CAN be mutable would make it even worse, but such a blanket statement might give people the wrong impression. Maybe something like:

> NuGet.org and NuGet client software are designed around a concept of package immutability where a .nupkg for a given package version will never change. The embedded files mentioned in this section support that concept. There are however third-party package feeds which allow varying degrees of package immutability. As such, there is now hard requirement of package immutability but due to client-side caching and other factors, it is recommended that authors and feed maintainers treat packages as immutable. 

Again, feel free to scrap it but this statement might confuse users or feeds implementors which have evidence to the contrary.

File: docs/api/implementation-guide.md
Comment (joelverhagen) [inline docs/api/implementation-guide.md:50]:
```suggestion
Packages that use a license expression, when represented in search and package metadata results, can have the `licenseUrl` set to the license expression, URL encoded, and appended to the end of https://licenses.nuget.org/. For more about NuGet.org's license content host and its protocol, see [https://learn.microsoft.com/en-us/nuget/nuget-org/licenses.nuget.org].
```

File: docs/api/implementation-guide.md
Comment (joelverhagen) [inline docs/api/implementation-guide.md:19]: Package types (returning them in the JSON and filtering by them) were added to search in 2019. Client does not use this but it was around 5.5.0 timeframe. .NET CLI does it it.

ID prefix reservation was implemented in mid to late 2017 and the "verified" boolean was added. 

SemVer 2.0.0 filtering was added in early 2017.

Not sure if you wanted those mentioned.

File: docs/api/implementation-guide.md
Comment (joelverhagen) [inline docs/api/implementation-guide.md:15]: Maybe have a section which is "Required endpoints" where you simply link to this section: https://learn.microsoft.com/en-us/nuget/api/overview#resources-and-schema and tell the reader to refer to the "required" column.

File: docs/api/implementation-guide.md
Comment (joelverhagen) [reply docs/api/implementation-guide.md:57]: We were holding off on documented depr+vuln in search until client actually does the implementation. We have flexibility to modify the protocol right now on NuGet.org since client doesn't use it (well, kind of... but yeah we have flexibility right now we would lose if we document the shape as is).

File: docs/api/implementation-guide.md
Comment (zivkan) [reply docs/api/implementation-guide.md:15]: Other than semver2, nothing seems essential (and given how old NuGet 3.6 is, supporting semver1 only clients seems like a non-issue any longer). Client can't assume that just because nuget.org implemented something that other servers have as well, so I really find it hard to imagine any change in the future that is truly required. 🤷

File: docs/api/implementation-guide.md
Comment (zivkan) [reply docs/api/implementation-guide.md:19]: This is utterly perfect, and exactly why I wanted as many people as possible to review.

> SemVer 2.0.0 filtering was added in early 2017.

I wasn't familiar with this before. While all my updates should be reviewed, please pay close attention to this one, because I largely made up everything I wrote based on my best guess and searching the existing resources docs for semver.

Comment (zivkan) [reply docs/api/implementation-guide.md]: @joelverhagen should we also add symbols packages info?

This doc has taken me so long, I'm losing motivation. Can I ask someone from the server team to add a section on symbols after I merge this? Especially given client's knowledge of symbols is basically limited to pdbs go in a different nupkg.

File: docs/api/implementation-guide.md
Comment (zivkan) [reply docs/api/implementation-guide.md]: I made a different change that hopefully explains it well enough.  I'm particularly concerned about too much information and being technically pedantic over practical, which will make it harder for developers of multi-ecosystem package repositories (nuget + npm + pip, etc) that aren't interested in NuGet's nuances.

File: docs/api/implementation-guide.md
Comment (zivkan) [reply docs/api/implementation-guide.md]: I've spent so many hours on this document now, I really need to focus on my other work, so I copy-pasted your text. To be perfectly honest, I think this is too long of an introduction. We all know that the more we write, the sooner we lose the attention of people who need to read it, so shorter is better. But I really want to get this merged and move on to other work, so maybe someone can open a PR to improve it after this PR is merged.

File: docs/api/implementation-guide.md
Comment (joelverhagen) [reply docs/api/implementation-guide.md:22]: Sounds good.

**Review by joelverhagen — APPROVED** [review]
Sorry for dragging this out with my verbosity. The doc is looking 100% shippable. :shipit:

File: docs/api/implementation-guide.md
Comment (joelverhagen) [reply docs/api/implementation-guide.md]: Probably, yes. We need to find some good links for the symserv HTTP protocol since the NuGet protocol is just ingress. Feel free to open a docs bug and put it on my plate.

**Pull request merged**
